app/scripts/migrations/048.js

```diff
@@ -87,5 +88,71 @@ function transformState (state = {}) {
     })
   }
 
+  // 6.  Convert address book keys from decimal to hex
+  const addressBook = state.AddressBookController?.addressBook || {}
+  Object.keys(addressBook).forEach((networkKey) => {
+    if ((/^\d+$/ui).test(networkKey)) {
+      const chainId = `0x${networkKey.toString(16)}`
+      updateChainIds(addressBook[networkKey], chainId)
+
+      if (addressBook[chainId]) {
+        mergeAddressBookKeys(addressBook, networkKey, chainId)
+      } else {
+        addressBook[chainId] = addressBook[networkKey]
+      }
+      delete addressBook[networkKey]
+    }
+  })
+
   return state
 }
+
+/**
+ * Merges the two given keys for the given address book in place.
+ *
+ * @returns {void}
+ */
+function mergeAddressBookKeys (addressBook, networkKey, chainIdKey) {
+  const networkKeyEntries = addressBook[networkKey] || {}
+  // For the new entries, start by copying the existing entries for the chainId
+  const newEntries = { ...addressBook[chainIdKey] }
+
+  // For each address of the old/network key entries
+  Object.keys(networkKeyEntries).forEach((address) => {
+    if (newEntries[address]) {
+      // Merge the corresponding chainId entry with the network entry
+      newEntries[address] = Object.keys(newEntries[address])
```

Comment:
Are we accounting for keys that exist in the old entry but not in the new one?

Reply:
Addressed in: 7ae9d4d